In [ ]:
import pandas as pd

# Load the data
file_path = '/content/Pizza.csv'
pizza_data = pd.read_csv(file_path)

# Display the first few rows of the dataset to understand its structure
pizza_data.head()

brand     id   mois   prot    fat   ash  sodium  carb   cal
0     A  14069  27.82  21.43  44.87  5.11    1.77  0.77  4.93
1     A  14053  28.49  21.26  43.89  5.34    1.79  1.02  4.84
2     A  14025  28.35  19.99  45.78  5.08    1.63  0.80  4.95
3     A  14016  30.55  20.15  43.13  4.79    1.61  1.38  4.74
4     A  14005  30.49  21.28  41.65  4.82    1.64  1.76  4.67

In [ ]:
# Question 1-(a)

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Perceptron
from sklearn.preprocessing import StandardScaler

# Extract feature columns
feature_columns = ['mois', 'prot', 'fat', 'ash', 'sodium', 'carb', 'cal']

# Function to train a perceptron and get the hyperplane coefficients
def train_perceptron_for_brands(data, brand1, brand2):
    # Filter data for the two brands
    filtered_data = data[data['brand'].isin([brand1, brand2])]

    # Convert brand labels to binary (brand1 = 0, brand2 = 1)
    filtered_data['label'] = filtered_data['brand'].apply(lambda x: 1 if x == brand2 else 0)

    # Split into features and labels
    X = filtered_data[feature_columns]
    y = filtered_data['label']

    # Standardize the features
    scaler = StandardScaler()
    X_scaled = scaler.fit_transform(X)

    # Train a perceptron model
    perceptron = Perceptron(max_iter=1000, tol=1e-3)
    perceptron.fit(X_scaled, y)

    # Get the hyperplane coefficients and the intercept
    coefficients = perceptron.coef_
    intercept = perceptron.intercept_

    return coefficients, intercept, scaler, perceptron

# Train perceptron for 'A' vs 'B'
coeff_AB, intercept_AB, scaler_AB, perceptron_AB = train_perceptron_for_brands(pizza_data, 'A', 'B')

# Train perceptron for 'A' vs 'C'
coeff_AC, intercept_AC, scaler_AC, perceptron_AC = train_perceptron_for_brands(pizza_data, 'A', 'C')

# Train perceptron for 'B' vs 'C'
coeff_BC, intercept_BC, scaler_BC, perceptron_BC = train_perceptron_for_brands(pizza_data, 'B', 'C')

coeff_AB, intercept_AB, coeff_AC, intercept_AC, coeff_BC, intercept_BC

<ipython-input-2-07170474f955>:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_data['label'] = filtered_data['brand'].apply(lambda x: 1 if x == brand2 else 0)
<ipython-input-2-07170474f955>:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_data['label'] = filtered_data['brand'].apply(lambda x: 1 if x == brand2 else 0)
<ipython-input-2-07170474f955>:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = val

(array([[ 1.13082672, -1.39272921, -0.77595361, -1.32480914, -1.27575277,
         -0.13057547, -0.94550352]]),
 array([-1.]),
 array([[ 0.86292306,  1.60642145, -1.02230547, -1.17379889, -1.12981306,
          0.07424423, -0.94682845]]),
 array([-1.]),
 array([[-1.76229001,  1.71740567, -1.21093847, -0.53704357, -2.0316657 ,
         -1.26047334, -0.62099529]]),
 array([0.]))

The Hyperplane coefficients and intercepts for the perceptron classifiers are as follows:-

1.) Brand A vs Brand B:

    

*   Coefficients = [1.13,-1.392,-0.775,-1.32,-1.275,-0.13,-0.945]
*   Intercept = -1

2.) Brand A vs Brand C:



*   Coefficients = [0.86,1.606,-1.02,-1.173,-1.298,0.07,-0.946]
*   Intercept = -1

3.) Brand B vs Brand C:

    

*   Coefficients = [-1.76,1.717,-1.2109,-0.537,-2.031,-1.2604,-0.62]
*   Intercept = 0



In [ ]:
# End of Question 1-(a)

In [ ]:
# Question 1-(b)

In [ ]:
import numpy as np

# Define the new instances
s1 = np.array([49.29, 24.82, 21.68, 2.76, 0.52, 1.47, 3.00])
s2 = np.array([30.95, 19.81, 42.28, 5.11, 1.67, 1.85, 4.67])
s3 = np.array([50.33, 13.28, 28.43, 3.58, 1.03, 4.38, 3.27])

# Function to compute margin for an instance
def compute_margin(instance, coef, intercept, scaler):
    # Standardize the instance based on the scaler used during training
    instance_scaled = scaler.transform([instance])

    # Compute the margin (dot product of coefficients and scaled features + intercept)
    margin = np.dot(coef, instance_scaled.T) + intercept
    return margin[0][0]

# Compute margins for each instance for all classifiers
margins_s1 = {
    'A vs B': compute_margin(s1, coeff_AB, intercept_AB, scaler_AB),
    'A vs C': compute_margin(s1, coeff_AC, intercept_AC, scaler_AC),
    'B vs C': compute_margin(s1, coeff_BC, intercept_BC, scaler_BC)
}

margins_s2 = {
    'A vs B': compute_margin(s2, coeff_AB, intercept_AB, scaler_AB),
    'A vs C': compute_margin(s2, coeff_AC, intercept_AC, scaler_AC),
    'B vs C': compute_margin(s2, coeff_BC, intercept_BC, scaler_BC)
}

margins_s3 = {
    'A vs B': compute_margin(s3, coeff_AB, intercept_AB, scaler_AB),
    'A vs C': compute_margin(s3, coeff_AC, intercept_AC, scaler_AC),
    'B vs C': compute_margin(s3, coeff_BC, intercept_BC, scaler_BC)
}

margins_s1, margins_s2, margins_s3

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but 

({'A vs B': 4.343802588533223,
  'A vs C': 5.304897121600211,
  'B vs C': 7.527737243607775},
 {'A vs B': -7.544181574792768,
  'A vs C': -7.293344715785398,
  'B vs C': 1.5198798314942694},
 {'A vs B': 4.80589701740064,
  'A vs C': -3.0106039892209924,
  'B vs C': -7.200092400969122})

The margins for the classifiers for each of the provided instances is as follows:-

Instance s1 = (49.29,24.82,21.68,2.76,0.52,1.47,3.00):



*   A vs B = 4.343
*   A vs C = 5.304

*   B vs C = 7.527

Instance s2 = (30.95,19.81,42.28,5.11,1.67,1.85,4.67):



*   A vs B = -7.544
*   A vs C = -7.293

*   B vs C = 1.519

Instance s3 = (50.33,13.28,28.43,3.58,1.03,4.38,3.27):



*   A vs B = 4.805
*   A vs C = -3.010

*   B vs C = -7.2

So, based on the margin values, the instances are classified according to the highest positive margin in the comparison

So, now we wil apply the fusion rule to classify each of these instances

In [ ]:
# End of Question 1-(b)

In [ ]:
# Question 1-(c)

In [ ]:
# Function to apply the fusion rule based on margin comparisons
def classify_instance(margins):
    if margins['A vs B'] > 0 and margins['A vs C'] > 0:
        return 'A'
    elif margins['A vs B'] < 0 and margins['B vs C'] > 0:
        return 'B'
    else:
        return 'C'

# Apply the fusion rule to classify s1, s2, and s3
classification_s1 = classify_instance(margins_s1)
classification_s2 = classify_instance(margins_s2)
classification_s3 = classify_instance(margins_s3)

classification_s1, classification_s2, classification_s3

('A', 'B', 'C')

Thus, based on the computed margins and after applying the fusion rule for classification. The final classifications are as follows:-



*   Instance s1 is classified as 'Brand A'
*   Instance s2 is classified as 'Brand B'

*   Instance s3 is classified as 'Brand C'





In [ ]:
# End of Question 1-(c)